cryptoprotocol/src/main/java/de/upb/crypto/incentive/cryptoprotocol/model/Setup.java

```diff
@@ -0,0 +1,137 @@
+        AesPseudorandomFunction prf = new AesPseudorandomFunction(PRF_KEY_LENGTH);
+
+        // instantiate SPS-EQ scheme used in this instance of the incentive system
+        SPSEQSignatureScheme spsEq = new SPSEQSignatureScheme(new SPSEQPublicParameters(bg));
+        SignatureKeyPair<SPSEQVerificationKey, SPSEQSigningKey> spsEqKeyPair = spsEq.generateKeyPair(2);
+        // TODO: save key pair in public parameters
+
+        // wrap up all values
+        return new PublicParameters(bg, w, h7, prf, spsEq, spsEqKeyPair);
+    }
+
+    /**
+     * generates a user key pair from public parameters
+     * @param pp object representation of public parameters
+     * @return object representation of a user key pair (see 2020 inc sys paper)
+     */
+    public static UserKeyPair userKeyGen(PublicParameters pp)
+    {
+        // draw random exponent for the user secret key
+        Zn usedZn = pp.getBG().getZn(); // the remainder class ring used in this instance of the incentive system
+        ZnElement usk = usedZn.getUniformlyRandomElement(); // secret exponent
+
+        // generate the user key for the PRF used to generate pseudorandomness in the system
+        PrfKey betaUsr = pp.getPrf().generateKey(); // key the user uses to generate pseudorandomness
+
+        // compute user public key from secret exponent
+        GroupElement upkElem = pp.getW().pow(usk);
+
+        // wrap up values
+        UserSecretKey USK = new UserSecretKey(usk, betaUsr);
+        UserPublicKey upk = new UserPublicKey(upkElem);
+        return new UserKeyPair(upk, USK);
+    }
+
+
+    public static ProviderKeyPair providerKeyGen(PublicParameters pp)
+    {
+        // draw the dlogs of the first 6 bases used in the Pedersen commitment in the token
+        ZnElement[] q = new ZnElement[6];
+        for(int i=0; i < q.length; i++)
+        {
+            q[i] = pp.getBG().getZn().getUniformlyRandomElement();
+        }
+
+        // compute above first 6 bases
+        GroupElement[] h = new GroupElement[6];
+        GroupElement g1Generator = pp.getBG().getG1().getGenerator();
+        for(int i = 0; i < h.length; i++)
+        {
+            h[i] =g1Generator.pow(q[i]);
+        }
+
+        // generate PRF key for provider
+        PrfKey betaProv = pp.getPrf().generateKey();
+
+        // wrap up values
+
+        SPSEQVerificationKey spseqVerificationKey = pp.getSpsEq().getVerificationKey(
+                pp.getSpsEqKeyPair().getVerificationKey().getRepresentation()
+        );
+        SPSEQSigningKey spseqSigningKey = pp.getSpsEq().getSigningKey(
+                pp.getSpsEqKeyPair().getSigningKey().getRepresentation()
+        );
+        ProviderPublicKey pk = null;
+        ProviderSecretKey sk = null;
+        try // if Pedersen Exception occurs during key object creation: respective key is null
+        {
+            pk = new ProviderPublicKey(spseqVerificationKey, h);
+        }
+        catch(PedersenException e)
```

Comment:
There are two cases here:
(1) it's impossible that this exception happens, or (2) it's possible.

In case (1), why catch/throw any exception anyway?
In case (2) don't catch the exception and go on as if nothing bad had happened. This is a catastrophic failure 😉. If this actually happens, then the rest of the execution is doomed anyway (there's no public key, so stuff is bound to fail), so better to fail early. Meaning: Don't catch the exception. Instead, let it go through and, if the caller doesn't expect/catch this exception, let it the exception terminate the complete program, which will have to happen at the next access to the key, anyway).

Reply:
See above comment, exception handling removed